text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+
+## Ember's Two Resolution Systems
+
+Ember currently has two distinct resolution systems. The first, which I will call the "Module Resolver", is the one that governs how `import` in Javascript works today.  It's not exactly the same as the node_modules spec, but it's close:
+
+ - it already incorporates key features like the `./index.js` convention.
+ - it handles relative specifiers (starting with `..` or `.`) exactly the same
+ - it handles bare package names in a way that is at least inspired by NPM packages, although the details differ.
+
+This RFC is not about changing how Module Resolver works. (Aligning it more closely with the broader node_modules semantics is the subject of other ongoing work, and can be it's own RFC.) For this RFC, we take the existing semantics as a given.
+
+The second resolution system in Ember apps is the one that locates components, helpers, routes, controllers, services, and other Ember-specifier object types. I will refer to this as the "Collection Resolver". The collection resolver's limitations are thoroughly described in [Module Unification, RFC 143](https://github.com/emberjs/rfcs/blob/master/text/0143-module-unification.md). In short, it has no locality (there's no such thing as relative lookup, so many things end up globally-scoped) and it has no notion of packages (all components, helpers, etc from addon packages go into one big global namespace soup).
+
+This RFC is not about deprecating or dropping the Collection Resolver. Doing so properly would require both a long period of transition, and alternative designs for doing some of the things that the Collection Resolver can do that the Module Resolver should not. For example, the Collection Resolver implements dependency injection, which is _not_ the same semantics as module imports. DI and modules serve different purposes in large scale systems.
+
+
+## Module Resolving in Ember Templates
+
+Specific proposed rules:
+
+1. We add the ability to use Ember's existing `import` semantics (the Module Resolver) from templates. This is supported in any template, irrespective of any "classic vs pods vs MU" distinction. This new `import` takes precedence, but anything not explicitly imported is handled the classic way by the Collection Resolver.
+2. Components and helpers inside MU code (meaning under `/src`) can _only_ be resolved by the Module Resolver, not the Collection Resolver. (See Compatibility Plan below for how addons can be compatible with both worlds.)
+3. Importing components and helpers that are authored in non-MU code is not supported.
+
+## Frontmatter syntax
+
+Imports appear within a new frontmatter section at the beginning of a template:
+
+```hbs
+---
+import Button from './button';
+import { Select } from 'ember-power-select';
+---
+<Select />
+<Button />
+```
+
+The frontmatter section (delimited by `---`) uses Javascript syntax, but it is limited to only allow import statements.
```

Comment:
If this was delimited with `<script> ... </script>` then we'd get syntax highlighting for free in editors, and we're opening the door for single-file components then it would make sense that pretty much any JS could go in there?

Downside being that it may be more confusing to have special treatment for the first script tag in the document… 

Also would potentially open the door to questions around lexical scoping or imports if it was allowed elsewhere in the document too (as would be expected for any other tag):

```hbs
<div>
  <script>
    import Button from './foo/button'
  </script>
  <Button />
</div>

<div>
  <script>
    import Button from './bar/button'
  </script>
  <Button />
</div>
```

🤔

Reply:
And if we are going the `<script>` route, might as well do `<template>` too.

Reply:
While `<script>` seems more approachable at first I think it does add trouble like mentioned. It's not full JS and I think we'd get dinged for that and I could see it being really hard to teach and limit what people are doing.

Making `<script>` do special imports would be a MAJOR breaking change to how Ember treats HBS too since currently popping in a script tag "just works" as if you threw in a script tag in `innerHTML` for an element. Not saying that's a good practice to have, but it is currently possible in Ember templates...